cockroach/sqlalchemy/dialect.py

```diff
@@ -0,0 +1,127 @@
+                                               **kwargs)
+
+    def initialize(self, connection):
+        # Bypass PGDialect's initialize implementation, which looks at
+        # server_version_info and performs postgres-specific queries
+        # to detect certain features on the server. Set the attributes
+        # by hand and hope things don't change out from under us too
+        # often.
+        super(PGDialect, self).initialize(connection)
+        self.implicit_returning = True
+        self.supports_native_enum = False
+        self.supports_smallserial = False
+        self._backslash_escapes = False
+
+    def _get_server_version_info(self, conn):
+        # PGDialect expects a postgres server version number here,
+        # although we've overridden most of the places where it's
+        # used.
+        return (9, 5, 0)
+
+    def _get_default_schema_name(self, conn):
+        raise NotImplementedError()
+
+    ddl_compiler = DDLCompiler
+    type_compiler = TypeCompiler
+
+    def get_table_names(self, conn, schema=None, **kw):
+        return [row.Table for row in conn.execute("SHOW TABLES")]
+
+    def has_table(self, conn, table, schema=None):
+        for i in self.get_table_names(conn, schema=schema):
+            if i == table:
+                return True
+        return False
+
+    def get_columns(self, conn, table_name, schema=None, **kw):
+        res = []
+        # TODO(bdarnell): escape table name
+        for row in conn.execute("SHOW COLUMNS FROM \"%s\"" % table_name):
+            name, type_name, nullable, default = row
+            # TODO(bdarnell): when there are type parameters, attach
+            # them to the returned type object.
+            typ = _type_map[type_name.split('(')[0].lower()]
+            res.append(dict(
+                name=name,
+                type=typ,
+                nullable=nullable,
+                default=default,
+            ))
+        return res
+
+    def get_indexes(self, conn, table_name, schema=None, **kw):
+        # Maps names to a bool indicating whether the index is unique.
+        uniques = collections.OrderedDict()
+        columns = collections.defaultdict(list)
+        # TODO(bdarnell): escape table name
+        for row in conn.execute("SHOW INDEXES FROM \"%s\"" % table_name):
+            columns[row.Name].append(row.Column)
+            uniques[row.Name] = row.Unique
+        res = []
+        # Map over uniques because it preserves order.
+        for name in uniques:
+            res.append(dict(name=name, column_names=columns[name], unique=uniques[name]))
+        return res
+
+    def get_pk_constraint(self, conn, table_name, schema=None, **kw):
+        # The PK is always first in the index list; it may not always
+        # be named "primary".
+        pk = self.get_indexes(conn, table_name, schema=schema, **kw)[0]
+        res = dict(constrained_columns=pk["column_names"])
+        if pk["name"] != "primary":
+            res["name"] = pk["name"]
```

Comment:
I'm probably missing something, but why not always do this (i.e. remove the `if`)?